include/libint2/dfbs_generator.h

```diff
@@ -0,0 +1,323 @@
+            auto alpha2 = shell2.alpha[0];
+            auto l1 = shell1.contr[0].l;
+            auto l2 = shell2.contr[0].l;
+            auto prefactor = std::pow((gamma_function(L + 2.) * gamma_function(l1 + l2 + 1.5)) /
+                                      (gamma_function(l1 + l2 + 2.) * gamma_function(L + 1.5)),
+                                      2.);
+            return prefactor * (alpha1 + alpha2);
+        }
+
+        /// @brief creates a set of product functions from a set of primitive shells
+        /// @param primitive_shells set of primitive shells
+        std::vector<Shell> product_functions(const std::vector<Shell> &primitive_shells) {
+            std::vector<Shell> product_functions;
+            for (auto i = 0; i < primitive_shells.size(); ++i) {
+                for (auto j = 0; j <= i; ++j) {
+                    auto li = primitive_shells[i].contr[0].l;
+                    auto lj = primitive_shells[j].contr[0].l;
+                    for (auto L = std::abs(li - lj); L <= li + lj; L++) {
+                        auto alpha = libint2::svector<double>({alpha_eff(primitive_shells[i], primitive_shells[j], L)});
+                        libint2::svector<Shell::Contraction> contr_;
+                        Shell::Contraction contr1;
+                        contr1.l = L;
+                        contr1.pure = true; // libint2 needs solid harmonics for 2c2b integrals
+                        contr1.coeff = {1.0};
+                        contr_.push_back(contr1);
+                        assert(primitive_shells[i].O == primitive_shells[j].O);
+                        auto shell = Shell(alpha, contr_, primitive_shells[i].O);
+                        if (std::find(product_functions.begin(), product_functions.end(),
+                                      shell) == product_functions.end())
+                            product_functions.emplace_back(shell);
+                    }
+                }
+            }
+            return product_functions;
+        }
+
+        /// @brief creates a set of candidate product shells from a set of primitive shells
+        /// @param primitive_shells set of primitive shells
+        /// @return set of candidate product shells
+        std::vector<std::vector<Shell>> candidate_functions(const std::vector<std::vector<Shell>> &primitive_shells) {
+            std::vector<std::vector<Shell>> candidate_functions;
+            for (auto i = 0; i < primitive_shells.size(); ++i) {
+                candidate_functions.push_back(product_functions(primitive_shells[i]));
+            }
+            return candidate_functions;
+        }
+
+        /// @brief returns a hash map of shell indices to basis function indices
+        std::vector<size_t> map_shell_to_basis_function(const std::vector<libint2::Shell> &shells) {
+            std::vector<size_t> result;
+            result.reserve(shells.size());
+
+            size_t n = 0;
+            for (auto shell: shells) {
```

Comment:
Review ALL range-for loops. Should rarely need `auto` capture, most of the time want `auto&&`